t/grpc_server_example/main.go

```diff
@@ -0,0 +1,254 @@
+	if person != nil {
+		if person.GetName() != "" {
+			msg += fmt.Sprintf(", name: %v", person.GetName())
+		}
+		if person.GetAge() != 0 {
+			msg += fmt.Sprintf(", age: %v", person.GetAge())
+		}
+	}
+
+	return &pb.HelloReply{
+		Message: msg,
+		Items:   in.GetItems(),
+		Gender:  in.GetGender(),
+	}, nil
+}
+
+func (s *server) SayHelloAfterDelay(ctx context.Context, in *pb.HelloRequest) (*pb.HelloReply, error) {
+
+	select {
+	case <-time.After(1 * time.Second):
+		fmt.Println("overslept")
+	case <-ctx.Done():
+		errStr := ctx.Err().Error()
+		if ctx.Err() == context.DeadlineExceeded {
+			return nil, status.Error(codes.DeadlineExceeded, errStr)
+		}
+	}
+
+	time.Sleep(1 * time.Second)
+
+	log.Printf("Received: %v", in.Name)
+
+	return &pb.HelloReply{Message: "Hello delay " + in.Name}, nil
+}
+
+func (s *server) Plus(ctx context.Context, in *pb.PlusRequest) (*pb.PlusReply, error) {
+	log.Printf("Received: %v %v", in.A, in.B)
+	return &pb.PlusReply{Result: in.A + in.B}, nil
+}
+
+// SayHelloServerStream streams HelloReply back to the client.
+func (s *server) SayHelloServerStream(req *pb.HelloRequest, stream pb.Greeter_SayHelloServerStreamServer) error {
+	log.Printf("Received server side stream req: %v\n", req)
+
+	// Say Hello 5 times.
+	for i := 0; i < 5; i++ {
+		if err := stream.Send(&pb.HelloReply{
+			Message: fmt.Sprintf("Hello %s", req.Name),
+		}); err != nil {
+			return status.Errorf(codes.Unavailable, "Unable to stream request back to client: %v", err)
+		}
+	}
+	return nil
+}
+
+// SayHelloClientStream receives a stream of HelloRequest from a client.
+func (s *server) SayHelloClientStream(stream pb.Greeter_SayHelloClientStreamServer) error {
+	log.Println("SayHello client side streaming has been initiated.")
+	cache := ""
+	for {
+		req, err := stream.Recv()
+		if err == io.EOF {
+			return stream.SendAndClose(&pb.HelloReply{Message: cache})
+		}
+		if err != nil {
+			return status.Errorf(codes.Unavailable, "Failed to read client stream: %v", err)
+		}
+		cache = fmt.Sprintf("%sHello %s!", cache, req.Name)
+	}
+}
+
+// SayHelloBidirectionalStream establishes a bidirectional stream with the client.
+func (s *server) SayHelloBidirectionalStream(stream pb.Greeter_SayHelloBidirectionalStreamServer) error {
+	log.Println("SayHello bidirectional streaming has been initiated.")
+
+	for {
+		req, err := stream.Recv()
+		if err == io.EOF {
+			return stream.Send(&pb.HelloReply{Message: "stream ended"})
+		}
+		if err != nil {
+			return status.Errorf(codes.Unavailable, "Failed to read client stream: %v", err)
+		}
+
+		// A small 0.5 sec sleep
+		time.Sleep(500 * time.Millisecond)
+
+		if err := stream.Send(&pb.HelloReply{Message: fmt.Sprintf("Hello %s", req.Name)}); err != nil {
+			return status.Errorf(codes.Unknown, "Failed to stream response back to client: %v", err)
+		}
+	}
+}
+
+func (s *server) Run(ctx context.Context, in *pb.Request) (*pb.Response, error) {
+	return &pb.Response{Body: in.User.Name + " " + in.Body}, nil
+}
```

Comment:
Is there any utility of this RPC?

Reply:
It is used in https://github.com/apache/apisix/blob/6bb9feac824f0640e0e4ab799e37bbbf3adcdd96/t/plugin/grpc-transcode2.t#L264

Reply:
Oh, I see. Thanks.